From 98b00b736c1f48655fb12ee27a4285b113bfa96b Mon Sep 17 00:00:00 2001
From: Ran Shalit <rshalit@habana.ai>
Date: Mon, 07 Sep 2020 19:13:24 +0300
Subject: [PATCH] [SW-12754] gaudi: security: add dma protection

Add DMA protection using protection bits.
This patch is done as part of security access of
SOC registers.

Change-Id: I10e032d08d4b188a372ec785524baf47201e2584
Signed-off-by: Ran Shalit <rshalit@habana.ai>
Signed-off-by: Igor Grinberg <igrinberg@habana.ai>
---

diff --git a/gaudi/security.c b/gaudi/security.c
index 3e0f451..8e38782 100644
--- a/gaudi/security.c
+++ b/gaudi/security.c
@@ -28,6 +28,8 @@
 #include <nif_rtr_0_regs.h>
 #include <nif_rtr_ctrl_0_regs.h>
 #include <psoc_global_conf_regs.h>
+#include <dma_if_w_s_down_ch0_regs.h>
+#include <dma_if_w_s_regs.h>
 
 #include <cpu_if_masks.h>
 #include <psoc_global_conf_masks.h>
@@ -292,6 +294,151 @@
 
 	/* Protect the private and internal registers */
 	set_prot_bits_range_all(base, SNGL_REG_PROT_SIZE * 2);
+}
+
+static void preboot_secure_dma_if_down_regs(u32 base)
+{
+	u32 mask, prot_base = base;
+
+	/* offsets: 0 - 0x580 */
+	set_prot_bits_range_all(prot_base, SNGL_REG_PROT_SIZE * 11);
+
+	/* offsets: 0x580 - 0x600 */
+	prot_base += SNGL_REG_PROT_SIZE * 11;
+	mask = ~(PROT_BIT(mmDMA_IF_W_S_DOWN_CH0_RANGE_SEC_BASE_LOW_AW_0) &
+		 PROT_BIT(mmDMA_IF_W_S_DOWN_CH0_RANGE_SEC_BASE_LOW_AW_1) &
+		 PROT_BIT(mmDMA_IF_W_S_DOWN_CH0_RANGE_SEC_BASE_LOW_AW_2) &
+		 PROT_BIT(mmDMA_IF_W_S_DOWN_CH0_RANGE_SEC_BASE_LOW_AW_3) &
+		 PROT_BIT(mmDMA_IF_W_S_DOWN_CH0_RANGE_SEC_BASE_LOW_AW_4) &
+		 PROT_BIT(mmDMA_IF_W_S_DOWN_CH0_RANGE_SEC_BASE_LOW_AW_5) &
+		 PROT_BIT(mmDMA_IF_W_S_DOWN_CH0_RANGE_SEC_BASE_LOW_AW_6));
+	regmap_writel(mask, PRIV_PROT_REG(prot_base));
+
+	/*
+	 * offsets: 0x600 - 0x680 allowed access
+	 * offsets: 0x680 - 0x700 as following:
+	 */
+	prot_base += SNGL_REG_PROT_SIZE * 2;
+	mask = (PROT_BIT(mmDMA_IF_W_S_DOWN_CH0_RANGE_PRIV_BASE_LOW_AW_0) &
+		PROT_BIT(mmDMA_IF_W_S_DOWN_CH0_RANGE_PRIV_BASE_LOW_AW_1) &
+		PROT_BIT(mmDMA_IF_W_S_DOWN_CH0_RANGE_PRIV_BASE_LOW_AW_2) &
+		PROT_BIT(mmDMA_IF_W_S_DOWN_CH0_RANGE_PRIV_BASE_LOW_AW_3) &
+		PROT_BIT(mmDMA_IF_W_S_DOWN_CH0_RANGE_PRIV_BASE_LOW_AW_4) &
+		PROT_BIT(mmDMA_IF_W_S_DOWN_CH0_RANGE_PRIV_BASE_LOW_AW_5) &
+		PROT_BIT(mmDMA_IF_W_S_DOWN_CH0_RANGE_PRIV_BASE_LOW_AW_6));
+	regmap_writel(mask, PRIV_PROT_REG(prot_base));
+
+	/* offsets: 0x700 - 0x800 */
+	prot_base += SNGL_REG_PROT_SIZE;
+	set_prot_bits_range_all(prot_base, SNGL_REG_PROT_SIZE * 2);
+
+	/* offsets: 0x800 - 0x880 as following */
+	prot_base += SNGL_REG_PROT_SIZE * 2;
+	mask = ~(PROT_BIT(mmDMA_IF_W_S_DOWN_CH0_RANGE_SEC_BASE_LOW_AR_0) &
+		 PROT_BIT(mmDMA_IF_W_S_DOWN_CH0_RANGE_SEC_BASE_LOW_AR_1) &
+		 PROT_BIT(mmDMA_IF_W_S_DOWN_CH0_RANGE_SEC_BASE_LOW_AR_2) &
+		 PROT_BIT(mmDMA_IF_W_S_DOWN_CH0_RANGE_SEC_BASE_LOW_AR_3) &
+		 PROT_BIT(mmDMA_IF_W_S_DOWN_CH0_RANGE_SEC_BASE_LOW_AR_4) &
+		 PROT_BIT(mmDMA_IF_W_S_DOWN_CH0_RANGE_SEC_BASE_LOW_AR_5) &
+		 PROT_BIT(mmDMA_IF_W_S_DOWN_CH0_RANGE_SEC_BASE_LOW_AR_6) &
+		 PROT_BIT(mmDMA_IF_W_S_DOWN_CH0_RANGE_SEC_BASE_LOW_AR_7) &
+		 PROT_BIT(mmDMA_IF_W_S_DOWN_CH0_RANGE_SEC_BASE_LOW_AR_8) &
+		 PROT_BIT(mmDMA_IF_W_S_DOWN_CH0_RANGE_SEC_BASE_LOW_AR_9) &
+		 PROT_BIT(mmDMA_IF_W_S_DOWN_CH0_RANGE_SEC_BASE_LOW_AR_10) &
+		 PROT_BIT(mmDMA_IF_W_S_DOWN_CH0_RANGE_SEC_BASE_LOW_AR_11) &
+		 PROT_BIT(mmDMA_IF_W_S_DOWN_CH0_RANGE_SEC_BASE_LOW_AR_12) &
+		 PROT_BIT(mmDMA_IF_W_S_DOWN_CH0_RANGE_SEC_BASE_LOW_AR_13) &
+		 PROT_BIT(mmDMA_IF_W_S_DOWN_CH0_RANGE_SEC_BASE_LOW_AR_14) &
+		 PROT_BIT(mmDMA_IF_W_S_DOWN_CH0_RANGE_SEC_BASE_LOW_AR_15) &
+		 PROT_BIT(mmDMA_IF_W_S_DOWN_CH0_RANGE_SEC_BASE_HIGH_AR_0) &
+		 PROT_BIT(mmDMA_IF_W_S_DOWN_CH0_RANGE_SEC_BASE_HIGH_AR_1) &
+		 PROT_BIT(mmDMA_IF_W_S_DOWN_CH0_RANGE_SEC_BASE_HIGH_AR_2) &
+		 PROT_BIT(mmDMA_IF_W_S_DOWN_CH0_RANGE_SEC_BASE_HIGH_AR_3) &
+		 PROT_BIT(mmDMA_IF_W_S_DOWN_CH0_RANGE_SEC_BASE_HIGH_AR_4) &
+		 PROT_BIT(mmDMA_IF_W_S_DOWN_CH0_RANGE_SEC_BASE_HIGH_AR_5) &
+		 PROT_BIT(mmDMA_IF_W_S_DOWN_CH0_RANGE_SEC_BASE_HIGH_AR_6));
+	regmap_writel(mask, PRIV_PROT_REG(prot_base));
+
+	/* offsets: 0x880 - 0x900 allowed access */
+
+	/* offsets: 0x900 - 0x980 as following: */
+	prot_base += SNGL_REG_PROT_SIZE * 2;
+	mask = ~(PROT_BIT(mmDMA_IF_W_S_DOWN_CH0_RANGE_SEC_MASK_HIGH_AR_7) &
+		 PROT_BIT(mmDMA_IF_W_S_DOWN_CH0_RANGE_SEC_MASK_HIGH_AR_8) &
+		 PROT_BIT(mmDMA_IF_W_S_DOWN_CH0_RANGE_SEC_MASK_HIGH_AR_9) &
+		 PROT_BIT(mmDMA_IF_W_S_DOWN_CH0_RANGE_SEC_MASK_HIGH_AR_10) &
+		 PROT_BIT(mmDMA_IF_W_S_DOWN_CH0_RANGE_SEC_MASK_HIGH_AR_11) &
+		 PROT_BIT(mmDMA_IF_W_S_DOWN_CH0_RANGE_SEC_MASK_HIGH_AR_12) &
+		 PROT_BIT(mmDMA_IF_W_S_DOWN_CH0_RANGE_SEC_MASK_HIGH_AR_13) &
+		 PROT_BIT(mmDMA_IF_W_S_DOWN_CH0_RANGE_SEC_MASK_HIGH_AR_14) &
+		 PROT_BIT(mmDMA_IF_W_S_DOWN_CH0_RANGE_SEC_MASK_HIGH_AR_15));
+	regmap_writel(mask, PRIV_PROT_REG(prot_base));
+
+	/* offsets: 0x980 - 0xA00 */
+	prot_base += SNGL_REG_PROT_SIZE;
+	set_prot_bits_range_all(prot_base, SNGL_REG_PROT_SIZE);
+
+	/* offsets: 0xA00 - 0xA80 */
+	prot_base += SNGL_REG_PROT_SIZE;
+	mask = ~(PROT_BIT(mmDMA_IF_W_S_DOWN_CH0_RANGE_SEC_HIT_AR) &
+		 PROT_BIT(mmDMA_IF_W_S_DOWN_CH0_RANGE_SEC_HIT_AW));
+	regmap_writel(mask, PRIV_PROT_REG(prot_base));
+
+	/* offsets: 0xA80 - 0xF80 */
+	prot_base += SNGL_REG_PROT_SIZE;
+	set_prot_bits_range_all(prot_base, PROT_SIZE_MAX - (prot_base - base));
+}
+
+static void preboot_secure_dma_if_regs(u32 base)
+{
+	u32 mask, prot_base = base;
+
+	/* 0 - 0x300 */
+	set_prot_bits_range_all(prot_base, SNGL_REG_PROT_SIZE * 6);
+
+	/*
+	 * 0x300 - 0x400 allowed
+	 * 0x400 - 0x500 not allowed access
+	 */
+	prot_base += SNGL_REG_PROT_SIZE * 8;
+	set_prot_bits_range_all(prot_base, SNGL_REG_PROT_SIZE * 2);
+
+	/* 0x500 - 0x600 allowed */
+	/* 0x600 - 0x700 not allowed */
+	prot_base += SNGL_REG_PROT_SIZE * 4;
+	set_prot_bits_range_all(prot_base, SNGL_REG_PROT_SIZE * 2);
+
+	/* 0x700 - 0x780 as following: */
+	prot_base += SNGL_REG_PROT_SIZE * 2;
+	mask = ~(PROT_BIT(mmDMA_IF_W_S_DMA0_HIT_RPROT) &
+		 PROT_BIT(mmDMA_IF_W_S_DMA0_HIT_WPROT) &
+		 PROT_BIT(mmDMA_IF_W_S_DMA1_HIT_RPROT) &
+		 PROT_BIT(mmDMA_IF_W_S_DMA1_HIT_WPROT));
+	regmap_writel(mask, PRIV_PROT_REG(prot_base));
+
+	/* 0x780 - 0xF80 as following: */
+	prot_base += SNGL_REG_PROT_SIZE;
+	set_prot_bits_range_all(prot_base, PROT_SIZE_MAX - (prot_base - base));
+}
+
+#define DMA_DIST		0x20000
+#define DMA_ENGINES		4
+
+static void preboot_secure_dma(void)
+{
+	u32 i, base, prot_base;
+
+	for (i = 0; i < DMA_ENGINES; i++) {
+		base = i * DMA_DIST;
+		prot_base = base + (mmDMA_IF_W_S_BASE & ~HL_CFG_BASE);
+		preboot_secure_dma_if_regs(prot_base);
+		prot_base = base + (mmDMA_IF_W_S_DOWN_BASE & ~HL_CFG_BASE);
+		set_prot_bits_range_all(prot_base, PROT_SIZE_MAX);
+		prot_base = base + (mmDMA_IF_W_S_DOWN_CH0_BASE & ~HL_CFG_BASE);
+		preboot_secure_dma_if_down_regs(prot_base);
+		prot_base = base + (mmDMA_IF_W_S_DOWN_CH1_BASE & ~HL_CFG_BASE);
+		preboot_secure_dma_if_down_regs(prot_base);
+	}
 }
 
 static void preboot_secure_cpu(void)
@@ -987,6 +1134,7 @@
 	preboot_secure_nic();
 	preboot_secure_if();
 	preboot_secure_sram_cfg();
+	preboot_secure_dma();
 
 	preboot_secure_hbw();
 
